packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -1503,69 +1602,322 @@ class Navigator extends StatefulWidget {
+  initial, // we'll want to run install; a route created by onGenerateInitialRoutes or by the initial widget.pages
+  pushing, // we're waiting for the future from didPush to complete
+  idle, // route is being harmless
+  // routes that are but will not present:
+  pop, // we'll want to call didPop
+  remove, // we'll want to run didReplace/didRemove etc
+  // routes that are not and will not present:
+  popping, // we're waiting for the route to call finalizeRoute to switch to dispose
+  removing, // we are waiting for subsequent routes to be done animating, then will switch to dispose
+  dispose, // we will dispose the route momentarily
+  disposed, // we have disposed the route
+}
+
+typedef _RouteEntryPredicate = bool Function(_RouteEntry entry);
+
+class _RouteEntry {
+  _RouteEntry(
+    this.route, {
+      @required _RouteLifecycle initialState,
+    }) : assert(route != null),
+      assert(initialState != null),
+      assert(
+      initialState == _RouteLifecycle.initial ||
+        initialState == _RouteLifecycle.push ||
+        initialState == _RouteLifecycle.pushReplace ||
+        initialState == _RouteLifecycle.replace
+      ),
+      currentState = initialState; // ignore: prefer_initializing_formals
+
+  final Route<dynamic> route;
+
+  _RouteLifecycle currentState;
+  Route<dynamic> lastAnnouncedNextRoute; // last argument to Route.didChangeNext
+  Route<dynamic> lastAnnouncedPreviousRoute; // last argument to Route.didChangePrevious
+  Route<dynamic> lastAnnouncedPoppedNextRoute; // last argument to Route.didPopNext
+
+  void handleAddition({ @required NavigatorState navigator, @required bool isNewFirst, @required Route<dynamic> previous, @required Route<dynamic> previousPresent }) {
+    assert(currentState == _RouteLifecycle.initial || currentState == _RouteLifecycle.push || currentState == _RouteLifecycle.pushReplace || currentState == _RouteLifecycle.replace);
+    assert(navigator != null);
+    assert(navigator._debugLocked);
+    assert(route._navigator == null);
+    final _RouteLifecycle previousState = currentState;
+    route._navigator = navigator;
+    route.install(isInitialRoute: currentState == _RouteLifecycle.initial);
+    assert(route.overlayEntries.isNotEmpty);
+    if (currentState == _RouteLifecycle.initial || currentState == _RouteLifecycle.push || currentState == _RouteLifecycle.pushReplace) {
+      currentState = _RouteLifecycle.pushing;
+      route.didPush().whenCompleteOrCancel(() {
+        if (currentState == _RouteLifecycle.pushing) {
+          currentState = _RouteLifecycle.idle;
+          assert(!navigator._debugLocked);
+          assert(() { navigator._debugLocked = true; return true; }());
+          navigator._flushHistoryUpdates();
+          assert(() { navigator._debugLocked = false; return true; }());
+        }
+      });
+    } else {
+      assert(currentState == _RouteLifecycle.replace);
```

Comment:
The assert and followed by the if with the same condition doesn't make much sense. Are we sure that the state is always replace here? Then remove the if.